go/test/endtoend/vtgate/prefixfanout/main_test.go

```diff
@@ -0,0 +1,278 @@
+          "type": "cfc"
+		}
+	},
+    "tables": {
+        "t1": {
+            "column_vindexes": [
+                {
+                    "column": "c1",
+                    "name": "cfc"
+                }
+			],
+			"columns": [
+				{
+					"name": "c2",
+					"type": "VARCHAR"
+				}
+			]
+		}
+    }
+}`
+
+	sKsMD5     = `cfc_testing_md5`
+	sSchemaMD5 = `
+CREATE TABLE t2 (
+c1 VARCHAR(20) NOT NULL,
+c2 varchar(40) NOT NULL,
+PRIMARY KEY (c1)
+) ENGINE=Innodb;`
+
+	sVSchemaMD5 = `
+{
+    "sharded": true,
+    "vindexes": {
+   	"cfc_md5": {
+		  "type": "cfc",
+		  "params": {
+			  "hash": "md5",
+			  "offsets": "[2]"
+		  }
+		}
+	},
+    "tables": {
+      	"t2": {
+            "column_vindexes": [
+                {
+                    "column": "c1",
+                    "name": "cfc_md5"
+                }
+			],
+			"columns": [
+				{
+					"name": "c2",
+					"type": "VARCHAR"
+				}
+			]
+        }
+    }
+}`
+)
+
+func TestMain(m *testing.M) {
+	defer cluster.PanicHandler(nil)
+	flag.Parse()
+
+	exitCode := func() int {
+		clusterInstance = cluster.NewCluster(cell, hostname)
+		defer clusterInstance.Teardown()
+
+		// Start topo server
+		if err := clusterInstance.StartTopo(); err != nil {
+			return 1
+		}
+
+		// Start keyspace
+		sKeyspace := &cluster.Keyspace{
+			Name:      sKs,
+			SchemaSQL: sSchema,
+			VSchema:   sVSchema,
+		}
+		// cfc_testing
+		if err := clusterInstance.StartKeyspace(*sKeyspace, []string{"-41", "41-4180", "4180-42", "42-"}, 0, false); err != nil {
+			return 1
+		}
+		// cfc_testing_md5
+		if err := clusterInstance.StartKeyspace(
+			cluster.Keyspace{
+				Name:      sKsMD5,
+				SchemaSQL: sSchemaMD5,
+				VSchema:   sVSchemaMD5,
+			}, []string{"-c2", "c2-c20a80", "c20a80-d0", "d0-"}, 0, false); err != nil {
+			return 1
+		}
+
+		// Start vtgate
+		if err := clusterInstance.StartVtgate(); err != nil {
+			return 1
+		}
+
+		return m.Run()
+	}()
+	os.Exit(exitCode)
+}
+
+func TestCFCPrefixQueryNoHash(t *testing.T) {
+	defer cluster.PanicHandler(t)
+	ctx := context.Background()
+	vtParams := mysql.ConnParams{
+		Host: "localhost",
+		Port: clusterInstance.VtgateMySQLPort,
+	}
+	conn, err := mysql.Connect(ctx, &vtParams)
+	require.Nil(t, err)
+	defer conn.Close()
+
+	exec(t, conn, "delete from t1")
+	defer exec(t, conn, "delete from t1")
+	// prepare the sentinel rows, i.e. every shard stores a row begins with letter A.
+	// hex ascii code of 'A' is 41. For a given primary key, e.g. 'AA' here, it should
+	// only legally belong to a single shard. We insert into all shards with different
+	// `c2` value so that we can test if a query fans out to all or not. Based on the
+	// following shard layout only "41-4180", "4180-42" should serve the rows staring with 'A'.
+	shards := []string{"-41", "41-4180", "4180-42", "42-"}
+	for i, s := range shards {
+		exec(t, conn, fmt.Sprintf("use `%s:%s`", sKs, s))
+		exec(t, conn, fmt.Sprintf("insert into t1 values('AA', 'shard-%d')", i))
+	}
```

Comment:
I'm all for putting this into the right place. I have a test in go/vt/vtgate/engine for this. It uses dummy statement rather than real statement. The closest I can find to exercise the end-to-end functionality is executor_select_test.go however each tests there uses `createLegacyExecutorEnv` which is not parameterized for shards. Where do you suggest to put this end-to-end functional testing ? @harshit-gangal

Reply:
I think we can keep this test.